overlord/snapstate/snapmgr.go

```diff
@@ -775,11 +824,17 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	if err != nil {
 		return err
 	}
+	var hadCandidateInSeq bool
+	err = t.Get("had-candidate-in-seq", &hadCandidateInSeq)
+	if err != nil {
+		return err
+	}
 
 	// relinking of the old snap is done in the undo of unlink-current-snap
-
-	snapst.Candidate = snapst.Sequence[len(snapst.Sequence)-1]
-	snapst.Sequence = snapst.Sequence[:len(snapst.Sequence)-1]
+	snapst.Candidate = snapst.Sequence[snapst.findIndex(snapst.Current)]
+	if hadCandidateInSeq {
+		snapst.Sequence = snapst.Sequence[:len(snapst.Sequence)-1]
```

Comment:
I think Gustavo suggestion was to remove it by revision and not assume it was at the end